fabric-transactions-v1/src/main/java/net/fabricmc/fabric/impl/transaction/TransactionImpl.java

```diff
@@ -0,0 +1,194 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.transaction;
+
+import java.util.ArrayList;
+import java.util.ConcurrentModificationException;
+import java.util.IdentityHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+
+import net.minecraft.server.MinecraftServer;
+
+import net.fabricmc.fabric.api.transaction.v1.Transaction;
+import net.fabricmc.fabric.api.transaction.v1.TransactionContext;
+import net.fabricmc.fabric.api.transaction.v1.TransactionParticipant;
+import net.fabricmc.fabric.api.transaction.v1.TransactionParticipant.TransactionDelegate;
+
+public class TransactionImpl implements Transaction {
+
+	private final class ContextImpl implements TransactionContext {
+		private ContextImpl() {
+		}
+
+		@Override
+		public <T> void setState(T state) {
+			stateStorage.put(contextDelegate, state);
+		}
+
+		@Override
+		@SuppressWarnings("unchecked")
+		public <T> T getState() {
+			return (T) stateStorage.get(contextDelegate);
+		}
+
+		@Override
+		public boolean isCommited() {
+			return isCommited;
+		}
+	}
+
+	private final ContextImpl context = new ContextImpl();
+	private boolean isOpen = true;
+	private boolean isCommited = false;
+	private final IdentityHashMap<TransactionDelegate, Consumer<TransactionContext>> participants = new IdentityHashMap<>();
+	private final IdentityHashMap<TransactionDelegate, Object> stateStorage = new IdentityHashMap<>();
+	private TransactionDelegate contextDelegate;
+
+	private TransactionImpl() {
+	}
+
+	@Override
+	public void close() {
+		if (isOpen) {
+			rollback();
+		}
+	}
+
+	private void clear() {
+		participants.clear();
+		stateStorage.clear();
+		contextDelegate = null;
+		isOpen = false;
+		isCommited = false;
+	}
+
+	private void validate() {
+		if (!isOpen) {
+			throw new IllegalStateException("Encountered transaction operation for closed transaction.");
+		}
+
+		if (STACK.get(stackPointer) != this) {
+			throw new IndexOutOfBoundsException("Transaction operations must apply to most recent open transaction.");
+		}
+
+		if (!innerLock.isHeldByCurrentThread()) {
+			throw new ConcurrentModificationException("Attempt to modify transaction status from foreign thread");
+		}
+	}
+
+	@Override
+	public void rollback() {
+		close(false);
+	}
+
+	@Override
+	public void commit() {
+		close(true);
+	}
+
+	private void close(boolean isCommited) {
+		validate();
+		this.isCommited = isCommited;
+
+		participants.forEach((c, r) -> {
+			contextDelegate = c;
+			r.accept(context);
+		});
+
+		clear();
+
+		final boolean root = --stackPointer == -1;
+
+		innerLock.unlock();
+
+		// non-server threads have an additional lock we must release.
```

Comment:
I personally feel allowing other threads to touch the Transaction state is not a good idea tbf.

We could bind a TransactionManager to a Thread (one for the client and one for the server).

Reply:
Well, if it isn't enforced then there can be mayhem, but fundamentally I see no problem with limiting this to the server thread only.  Others (I forget who) have said multi-thread support is a must-have.

But allowing independent transactions on separate threads becomes problematic if there isn't complete isolation of *all* state between threads when a transaction is open on either because even if you synchronize both threads at that point, you're reading state that could be rolled back and you'd never know it.  

One could start to borrow from RDBMS here, adding per-object locks and structures to track transaction scope across threads, but you then have to deal with much more complication,  deadlocks, etc.  I explain this at some length in the header.